persist-cli/src/main/java/io/ballerina/persist/introspect/Introspector.java

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (c) 2024 WSO2 LLC. (http://www.wso2.com) All Rights Reserved.
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.ballerina.persist.introspect;
+
+import io.ballerina.persist.inflector.CaseConverter;
+import io.ballerina.persist.inflector.Pluralizer;
+import io.ballerina.persist.introspectiondto.SQLEnum;
+import io.ballerina.persist.introspectiondto.SQLForeignKey;
+import io.ballerina.persist.introspectiondto.SQLTable;
+import io.ballerina.persist.models.Entity;
+import io.ballerina.persist.models.EntityField;
+import io.ballerina.persist.models.Enum;
+import io.ballerina.persist.models.EnumMember;
+import io.ballerina.persist.models.Index;
+import io.ballerina.persist.models.Module;
+import io.ballerina.persist.models.Relation;
+import io.ballerina.persist.models.SQLType;
+import io.ballerina.persist.utils.ScriptRunner;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.lang.Integer.parseInt;
+
+
+
+
+/**
+ * Database Introspector class.
+ *
+ *
+ */
+public abstract class Introspector {
+
+    protected final Connection connection;
+    protected String databaseName;
+    protected abstract String getTablesQuery();
+    protected abstract String getColumnsQuery(String tableName);
+    protected abstract String getIndexesQuery(String tableName);
+    protected abstract String getForeignKeysQuery(String tableName);
+    protected abstract String getEnumsQuery();
+
+    private List<SQLTable> tables;
+
+    private List<SQLEnum> sqlEnums;
+    private final Module.Builder moduleBuilder;
+
+    private final Map<String, Entity> entityMap;
+
+    private final List<SQLForeignKey> sqlForeignKeys;
+
+    public Introspector(Connection connection, String databaseName) {
+        this.connection = connection;
+        this.databaseName = databaseName;
+        this.tables = new ArrayList<>();
+        this.entityMap = new HashMap<>();
+        this.sqlForeignKeys = new ArrayList<>();
+        this.sqlEnums = new ArrayList<>();
+        this.moduleBuilder = Module.newBuilder("db");
+    }
+
+    public Module introspectDatabase() throws SQLException {
```

Comment:
`SQLException` is never thrown here right?

Reply:
It is thrown by the ScriptRunner methods.